libfreerdp/core/gcc.c

```diff
@@ -829,49 +853,139 @@ BOOL gcc_write_user_data_header(wStream* s, UINT16 type, UINT16 length)
+	     RNS_UD_CS_VALID_CONNECTION_TYPE | RNS_UD_CS_SUPPORT_MONITOR_LAYOUT_PDU |
+	     RNS_UD_CS_SUPPORT_NETCHAR_AUTODETECT | RNS_UD_CS_SUPPORT_DYNVC_GFX_PROTOCOL |
+	     RNS_UD_CS_SUPPORT_DYNAMIC_TIME_ZONE |
+	     RNS_UD_CS_SUPPORT_HEARTBEAT_PDU); // unsupported RNS_UD_CS_SUPPORT_SKIP_CHANNELJOIN
+	const UINT32 filtered = flags & mask;
+	const UINT32 unknown = flags & ~mask;
+	if (unknown != 0)
+	{
+		char buffer[256] = { 0 };
+		WLog_WARN(TAG,
+		          "(TS_UD_CS_CORE)::EarlyCapabilityFlags [0x%08" PRIx32 " & 0x%08" PRIx32
+		          " --> 0x%08" PRIx32 "] filtering %s, feature not implemented",
+		          flags, ~mask, unknown,
+		          rdp_early_client_caps_string(unknown, buffer, sizeof(buffer)));
+	}
+	return filtered;
+}
+
+static UINT16 earlyClientCapsFromSettings(const rdpSettings* settings)
+{
+	UINT32 earlyCapabilityFlags = 0;
 
 	WINPR_ASSERT(settings);
 	if (settings->SupportErrorInfoPdu)
 		earlyCapabilityFlags |= RNS_UD_CS_SUPPORT_ERRINFO_PDU;
 
+	if (freerdp_settings_get_uint32(settings, FreeRDP_ColorDepth) == 32)
+		earlyCapabilityFlags |= RNS_UD_CS_WANT_32BPP_SESSION;
+
+	if (settings->SupportStatusInfoPdu)
+		earlyCapabilityFlags |= RNS_UD_CS_SUPPORT_STATUSINFO_PDU;
+
 	if (settings->ConnectionType)
 		earlyCapabilityFlags |= RNS_UD_CS_VALID_CONNECTION_TYPE;
 
-	if (freerdp_settings_get_uint32(settings, FreeRDP_ColorDepth) == 32)
-		earlyCapabilityFlags |= RNS_UD_CS_WANT_32BPP_SESSION;
+	if (settings->SupportMonitorLayoutPdu)
+		earlyCapabilityFlags |= RNS_UD_CS_SUPPORT_MONITOR_LAYOUT_PDU;
 
 	if (freerdp_settings_get_bool(settings, FreeRDP_NetworkAutoDetect))
 		earlyCapabilityFlags |= RNS_UD_CS_SUPPORT_NETCHAR_AUTODETECT;
 
-	if (settings->SupportHeartbeatPdu)
-		earlyCapabilityFlags |= RNS_UD_CS_SUPPORT_HEARTBEAT_PDU;
-
 	if (settings->SupportGraphicsPipeline)
 		earlyCapabilityFlags |= RNS_UD_CS_SUPPORT_DYNVC_GFX_PROTOCOL;
 
 	if (settings->SupportDynamicTimeZone)
 		earlyCapabilityFlags |= RNS_UD_CS_SUPPORT_DYNAMIC_TIME_ZONE;
 
-	if (settings->SupportMonitorLayoutPdu)
-		earlyCapabilityFlags |= RNS_UD_CS_SUPPORT_MONITOR_LAYOUT_PDU;
+	if (settings->SupportHeartbeatPdu)
+		earlyCapabilityFlags |= RNS_UD_CS_SUPPORT_HEARTBEAT_PDU;
 
-	if (settings->SupportStatusInfoPdu)
-		earlyCapabilityFlags |= RNS_UD_CS_SUPPORT_STATUSINFO_PDU;
+	if (settings->SupportAsymetricKeys)
+		earlyCapabilityFlags |= RNS_UD_CS_STRONG_ASYMMETRIC_KEYS;
+
+	if (settings->SupportSkipChannelJoin)
+		earlyCapabilityFlags |= RNS_UD_CS_SUPPORT_SKIP_CHANNELJOIN;
 
-	return earlyCapabilityFlags;
+	return filterAndLogEarlyClientCapabilityFlags(earlyCapabilityFlags);
 }
 
 static BOOL updateEarlyClientCaps(rdpSettings* settings, UINT32 earlyCapabilityFlags,
                                   UINT32 connectionType)
 {
 	WINPR_ASSERT(settings);
 
+	if (settings->SupportErrorInfoPdu)
+		settings->SupportErrorInfoPdu =
+		    (earlyCapabilityFlags & RNS_UD_CS_SUPPORT_ERRINFO_PDU) ? TRUE : FALSE;
+
+#if 0
```

Comment:
If not needed the code should be removed (or even better replaced with a comment why we do not set this flag here)

Reply:
ok, replaced with a comment why that flag is not handled in this function.